Add OpenLDAP Tutorial (#334)
stackabletech/documentation: modules/tutorials/pages/authentication_with_openldap.adoc

= Authentication with OpenLDAP

The Stackable platform supports user authentication with LDAP in multiple products. This page guides you through setting up the configuration necessary to use an exising LDAP installation with Stackable supported products.

Prerequisites:

* a k8s cluster
* xref:stackablectl::index.adoc[] installed
* basic knowledge of how to create resources in Kubernetes (i.e. `kubectl apply -f <filename>.yaml`) and inspect them (`kubectl get` or a tool like `k9s`)

== Setup

> Before configuring LDAP you need to set up some services to configure, as well as an LDAP to use for authentication. This tutorial is about learning how to configure an LDAP, so there won't be detailed instructions on how to set up all of this, instead we will use xref:stackablectl::commands/stack.adoc[stackablectl Stacks]. 

Reviewer: Nit: we switch a bit between LDAP (the protocol) and **_an_** LDAP (installation): I would make that little more explicit e.g.
```suggestion
Before configuring LDAP you need to set up some services to configure, as well as an LDAP server to use for authentication. This tutorial is about learning how to configure LDAP, so there won't be detailed instructions on how to set up all of this, instead we will use xref:stackablectl::commands/stack.adoc[stackablectl Stacks]. 
```

Author: good point! I think it's important to be precise with language especially in tutorials

=== Superset and Trino Stack

This command will install the latest Stackable release for you, and set up the `trino-superset-s3` Stack. It contains a Superset instance connected to a Trino instance, and all their dependencies (Minio S3, PostgreSQL). Later in this guide, a Superset and Trino instance will be configured to use LDAP.

[source,bash]
include::example$ldap-auth/10-install-base.sh[tag=stackablectl-install]

This command will take a few minutes to complete.

NOTE: The stack installed here is used in the xref:stackablectl::demos/trino-taxi-data.adoc[`trino-tax-data` demo]. Click the link to learn more.

[#setup_login]
==== Inspect

> Before starting with adding configuration to your Stackable cluster, inspect what the command above has set up in your Kubernetes cluster.

Reviewer: ```suggestion
Before starting to add configuration to your Stackable cluster, inspect what the command above has set up in your Kubernetes cluster.
```

Use `stackablectl svc list` to find the endpoints of Superset and Trino and open their web interfaces in the Browser.

You can log into Superset with user _admin_ and password _adminadmin_, and into Trino with user _admin_ and password _adminadmin_.

> These are the current users defined in Supersets and Trinos internal user management. Later you will see that these users cannot be used for authentication anymore after LDAP authentication has been enabled.

Reviewer: ```suggestion
These are the current users defined in Superset's and Trino's internal user management. Later you will see that these users cannot be used for authentication anymore after LDAP authentication has been enabled.
```

First, create a secret that contains the LDAP bind credentials which products can use to authenticate with LDAP:

[source,yaml]
include::example$ldap-auth/bind-credentials-secret.yaml[]

<1> The annotation which SecretClass this secret belongs to
<2> The LDAP bind user that was provided by the `tutorial-ldap` Stack
<3> The corresponding password

// [source,bash]
// include::example$ldap-auth/30-install-openldap.sh[tag=apply-bind-credentials-secret]

Notice the SecretClass annotation. Create the SecretClass next:

[source,yaml]
include::example$ldap-auth/bind-credentials-secretclass.yaml[]

// [source,bash]
// include::example$ldap-auth/30-install-openldap.sh[tag=apply-credentials-secretclass]

This level of indirection is necessary, because the AuthenticationClass is cluster-scoped but Secrets are not.

Now you can create the AuthenticationClass `openldap` which references the SecretClass:

[source,yaml]
include::example$ldap-auth/ldap-authenticationclass.yaml[]

> <1> The name of the AuthenticationClass, needs to be referenced later

Reviewer: ```suggestion
<1> The name of the AuthenticationClass, which needs to be referenced later
```

[source,bash]
include::example$ldap-auth/40-modify-superset.sh[tag=apply-superset-cluster]

Connect to superset as before, and try logging in again with username _admin_ and password _adminadmin_, Superset will not accept these credentials anymore. You now have to use LDAP credentials to log in. The OpenLDAP you installed earlier comes with two users, _alice_ (password _alice_) and _bob_ (password _bob_). Log in with any of these users and Superset will accept.

[#trino]
=== Add LDAP configuration to Trino

Trino is configured very similarly to Superset.

Fetch the existing TrinoCluster defintion from the Kubernetes API server and save it into a `trino.yaml` file:

[source,bash]
include::example$ldap-auth/60-modify-trino.sh[tag=get-yaml]

.The `trino.yaml` file should look similar to this
[%collapsible]
====
[source,yaml]
----
---
apiVersion: trino.stackable.tech/v1alpha1
kind: TrinoCluster
metadata:
  name: trino
  ...
spec:
  version: 396-stackable0.1.0
  authentication:
    method:
      multiUser:
        ...
  opa:
    configMapName: opa
    package: trino
  catalogLabelSelector:
    ...
  workers:
    ...
  coordinators:
    ...
  ...
----
====

> You can now delete the TrinoCluster, we will recreate it later with the new configuration:

Reviewer: ```suggestion
You can now delete the TrinoCluster. We will recreate it later with the new configuration:
```
